src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -1047,24 +775,25 @@ impl<'a, 'tcx> Checker<'a, 'tcx> {
         // Collect all the temps we need to promote.
         let mut promoted_temps = BitSet::new_empty(self.temp_promotion_state.len());
 
-        // HACK(eddyb) don't try to validate promotion candidates if any
-        // parts of the control-flow graph were skipped due to an error.
-        let promotion_candidates = if has_controlflow_error {
-            let unleash_miri = self
-                .tcx
-                .sess
-                .opts
-                .debugging_opts
-                .unleash_the_miri_inside_of_you;
-            if !unleash_miri {
-                self.tcx.sess.delay_span_bug(
-                    body.span,
-                    "check_const: expected control-flow error(s)",
-                );
-            }
-            self.promotion_candidates.clone()
+        // HACK: if parts of the control-flow graph were skipped due to an error, don't try to
+        // promote anything, since that can cause errors in a `const` if e.g. rvalue static
+        // promotion is attempted within a loop body.
+        let unleash_miri = self.tcx.sess.opts.debugging_opts.unleash_the_miri_inside_of_you;
+        let promotion_candidates = if has_controlflow_error && !unleash_miri {
```

Comment:
I was going to say that this isn't needed (since IIRC the reason I added it was because there would be mismatches since `qualify_consts` doesn't see as many candidates as there are in the body, because it stops looking at BBs) but you did replace my `HACK` comment so I'm guessing you hit some other issue?

Actually, I'm still not sure this check is needed, promotion within `fn` bodies works just fine without it, so what gives?

Reply:
We can't do rvalue-static promotion in a loop with storage dead removal, it results in broken MIR and ultimately an ICE as the stack slot for the temp is never cleaned up.

```rust
const X: () = loop {
    let x = &4;
};
```

Reply:
Actually, there's no ICE on the test that hits this (it's been a few days since I ran them). There's just a MIRI error:

```
+	error[E0384]: cannot assign twice to immutable variable `_`
+	  --> $DIR/issue-52475.rs:10:18
+	   |
+	LL |             x = &0; // Materialize a new AllocId
+	   |                  ^ cannot assign twice to immutable variable
+	

```